File: Solutions.java/Graph/1579. Remove Max Number of Edges to Keep Graph Fully Traversable.java
/// <summary>
/// 1579. Remove Max Number of Edges to Keep Graph Fully Traversable
/// https://leetcode.com/problems/remove-max-number-of-edges-to-keep-graph-fully-traversable
/// 
/// Difficulty Hard
/// Acceptance 71.0%
/// 
/// Union Find
/// Graph
/// </summary>
class Solution {
	public int maxNumEdgesToRemove(int n, int[][] edges) {
		var mesh = new ArrayList<HashSet<Integer>>();
		var edges1 = new ArrayList<int[]>();
		var edges2 = new ArrayList<int[]>();
		var lack = n - 1;

		for (var edge : edges) {
			var type = edge[0];
			if (type == 1) {
				edges1.add(new int[] {edge[1], edge[2]});
			} else if (type == 2) {
				edges2.add(new int[] {edge[1], edge[2]});
			} else {
				lack -= proceed(mesh, edge[1], edge[2]);
				if (lack == 0) {
					return edges.length - n + 1;
				}
			}
		}

		var meshClone = new ArrayList<>(mesh.stream().map(h -> (HashSet<Integer>) h.clone()).toList());
		if (checkMesh(meshClone, edges1, lack) > 0 || checkMesh(mesh, edges2, lack) > 0) {
			return -1;
		}

        return edges.length - n + 1 - lack;
    }

	private static int checkMesh(ArrayList<HashSet<Integer>> mesh, ArrayList<int[]> edges, int lack) {
		for (var edge : edges) {
			lack -= proceed(mesh, edge[0], edge[1]);
			if (lack == 0) {
				break;
			}
		}
		return lack;
	}

	private static int proceed(ArrayList<HashSet<Integer>> mesh, int edge0, int edge1) {
		var subs = find(mesh, edge0, edge1);
		if (subs[0] == null && subs[1] == null) {
			var set = new HashSet<Integer>();
			set.add(edge0);
			set.add(edge1);
			mesh.add(set);
			return 1;
		} else if (subs[0] == null && subs[1] != null) {
			subs[1].add(edge0);
			return 1;
		} else if (subs[0] != null && subs[1] == null) {
			subs[0].add(edge1);
			return 1;
		} else if (subs[0] == subs[1]) {
			return 0;
		} else {
			subs[0].addAll(subs[1]);
			mesh.remove(subs[1]);
			return 1;
		}
	}

	private static HashSet<Integer>[] find(ArrayList<HashSet<Integer>> mesh, int edge0, int edge1) {
		var result = new HashSet[2];
		for (var m : mesh) {
			if (m.contains(edge0)) {
				result[0] = m;
			}
			if (m.contains(edge1)) {
				result[1] = m;
			}

			if (result[0] != null && result[1] != null) {
				break;
			}
		}

		return result;
	}
}

File: Solutions.cs/Union Find/1579. Remove Max Number of Edges to Keep Graph Fully Traversable.cs
namespace Problem1579
{

  /// <summary>
  /// 1579. Remove Max Number of Edges to Keep Graph Fully Traversable
  /// https://leetcode.com/problems/remove-max-number-of-edges-to-keep-graph-fully-traversable
  /// 
  /// Difficulty Hard
  /// Acceptance 71.0%
  /// 
  /// Union Find
  /// Graph
  /// </summary>
  public class Solution
  {
    public int MaxNumEdgesToRemove(int n, int[][] edges)
    {
      if (edges.Length < n - 1)
        return -1;

      var graphUnion = new UnionFind(n);
      var edges1 = new List<(int A, int B)>();
      var edges2 = new List<(int A, int B)>();
      var lack = n - 1;

      foreach (var edge in edges)
      {
        var type = edge[0];
        if (type == 1)
        {
          edges1.Add((edge[1], edge[2]));
        }
        else if (type == 2)
        {
          edges2.Add((edge[1], edge[2]));
        }
        else
        {
          if (graphUnion.TryUnite(edge[1], edge[2]))
          {
            lack--;

            if (lack == 0)
              return edges.Length - n + 1;
          }
        }
      }

      var graphUnionClone = new UnionFind(graphUnion);
      if (IsBad(graphUnionClone, edges1, lack) ||
        IsBad(graphUnion, edges2, lack))
        return -1;

      return edges.Length - n + 1 - lack;
    }

    private static bool IsBad(UnionFind graphUnion, List<(int A, int B)> edges, int lack)
    {
      if (edges.Count < lack)
        return true;

      foreach (var edge in edges)
      {
        if (graphUnion.TryUnite(edge.A, edge.B))
          lack--;

        if (lack == 0)
          return false;
      }
      return lack > 0;
    }

    private class UnionFind
    {
      private readonly int Size;
      private readonly int[] Parents;
      private readonly int[] ChildrenCount;

      public int Groups { get; private set; }

      public UnionFind(int size)
      {
        Size = size;
        Parents = Enumerable.Range(0, Size).ToArray();
        ChildrenCount = Enumerable.Range(0, Size).Select(i => 1).ToArray();
        Groups = size;
      }

      public UnionFind(UnionFind other)
      {
        Size = other.Size;
        Parents = other.Parents.ToArray();
        ChildrenCount = other.Parents.ToArray();
        Groups = other.Groups;
      }

      public bool TryUnite(int node0, int node1)
      {
        node0--;
        node1--;

        if (node0 == node1)
          return false;

        node0 = FindSubroot(node0);
        node1 = FindSubroot(node1);

        if (node0 == node1)
          return false;

        if (ChildrenCount[node0] < ChildrenCount[node1])
        {
          Parents[node0] = node1;
          ChildrenCount[node1] += ChildrenCount[node0];
        }
        else
        {
          Parents[node1] = node0;
          ChildrenCount[node0] += ChildrenCount[node1];
        }

        Groups--;

        return true;
      }

      private int FindSubroot(int node)
      {
        while (Parents[node] != node)
          node = Parents[node];

        return node;
      }
    }
  }
}
